src/autoscaling/snapshotter.js

```diff
@@ -0,0 +1,267 @@
+     * @ignore
+     */
+    getMemorySample(sampleDurationMillis) {
+        return this._getSample(this.memorySnapshots, sampleDurationMillis);
+    }
+
+    /**
+     * Returns a sample of latest event loop snapshots, with the size of the sample defined
+     * by the sampleDurationMillis parameter. If omitted, it returns a full snapshot history.
+     * @param {Number} [sampleDurationMillis]
+     * @return {Array} sample
+     * @ignore
+     */
+    getEventLoopSample(sampleDurationMillis) {
+        return this._getSample(this.eventLoopSnapshots, sampleDurationMillis);
+    }
+
+    /**
+     * Returns a sample of latest CPU snapshots, with the size of the sample defined
+     * by the sampleDurationMillis parameter. If omitted, it returns a full snapshot history.
+     * @param {Number} [sampleDurationMillis]
+     * @return {Array} sample
+     * @ignore
+     */
+    getCpuSample(sampleDurationMillis) {
+        return this._getSample(this.cpuSnapshots, sampleDurationMillis);
+    }
+
+    /**
+     * Finds the latest snapshots by sampleDurationMillis in the provided array.
+     * @param {Array} snapshots
+     * @param {Number} [sampleDurationMillis]
+     * @return {Array} sample
+     * @ignore
+     */
+    _getSample(snapshots, sampleDurationMillis) { // eslint-disable-line class-methods-use-this
+        if (!sampleDurationMillis) return snapshots;
+
+        const sample = [];
+        let idx = snapshots.length;
+        if (!idx) return sample;
+
+        const latestTime = snapshots[idx - 1].createdAt;
+        while (idx--) {
+            const snapshot = snapshots[idx];
+            if (latestTime - snapshot.createdAt <= sampleDurationMillis) sample.unshift(snapshot);
+            else break;
+        }
+        return sample;
+    }
+
+    /**
+     * Creates a snapshot of current memory usage.
+     * @param {Function} intervalCallback
+     * @return {Promise}
+     * @ignore
+     */
+    async _snapshotMemory(intervalCallback) {
+        try {
+            const now = new Date();
+            const memInfo = await getMemoryInfo();
+            const { mainProcessBytes, childProcessesBytes } = memInfo;
+            this._pruneSnapshots(this.memorySnapshots, now);
+
+            const usedBytes = mainProcessBytes + childProcessesBytes;
+            const snapshot = Object.assign(memInfo, {
+                createdAt: now,
+                isOverloaded: usedBytes / this.maxMemoryBytes > this.maxUsedMemoryRatio,
+            });
+
+            this.memorySnapshots.push(snapshot);
+        } catch (err) {
+            log.exception(err, 'Snapshotter: Memory snapshot failed.');
+        } finally {
+            intervalCallback();
+        }
+    }
+
+    /**
+     * Creates a snapshot of current event loop delay.
+     * @param {Function} intervalCallback
+     * @return {Promise}
+     * @ignore
+     */
+    _snapshotEventLoop(intervalCallback) {
+        try {
+            const now = new Date();
```

Comment:
`blocked` package is using high-resolution timer for this (`process.hrtime()`), perhaps they know why - see https://github.com/tj/node-blocked/blob/master/index.js

Reply:
I did some research and found no evidence that `process.hrtime()` is faster than `new Date()` so I'm using Date for convenience, since we don't need the extra resolution. Also Date serializes to a date string, which is what @mtrunkat wanted.

Reply:
ok

Reply:
Actually, there is a good reason why use `process.hrtime` instead (https://nodejs.org/api/process.html#process_process_hrtime_time): "These times are relative to an arbitrary time in the past, and not related to the time of day and therefore not subject to clock drift". This means it works even if user or NTPD updates system time

Reply:
Yeah, but then the timestamps will not be serializable to ISOString easily and therefore not in line with our other `createdAt` properties. @mtrunkat What do you think about this?